templates/telemetry_derived/error_aggregates/query.sql

```diff
@@ -0,0 +1,196 @@
+crash_ping_data AS (
+  SELECT
+    submission_timestamp,
+    normalized_channel AS channel,
+    environment.build.version,
+    application.display_version,
+    environment.build.build_id,
+    metadata.uri.app_name,
+    IF(metadata.uri.app_name = 'Fennec'
+       AND environment.system.os.name = 'Linux',
+       'Android',
+       environment.system.os.name) AS os,
+    environment.system.os.version AS os_version,
+    environment.build.architecture,
+    normalized_country_code AS country,
+    IF(
+      payload.process_type = 'main'
+      OR payload.process_type IS NULL,
+      1,
+      0
+    ) AS main_crash,
+    IF(
+      REGEXP_CONTAINS(payload.process_type, 'content'),
+      1,
+      0
+    ) AS content_crash,
+    IF(payload.metadata.startup_crash = '1', 1, 0) AS startup_crash,
+    IF(
+      REGEXP_CONTAINS(payload.metadata.ipc_channel_error, 'ShutDownKill'),
+      1,
+      0
+    ) AS content_shutdown_crash,
+    -- 0 for values retrieved from main/core pings
+    0 AS usage_hours,
+    0 AS gpu_crashes,
+    0 AS plugin_crashes,
+    0 AS gmplugin_crashes
+  FROM
+    crash_pings
+),
+-- Get desktop usage hours and gpu, plugin, and gmplugin crashes from main pings
+main_ping_data AS (
+  SELECT
+    submission_timestamp,
+    normalized_channel AS channel,
+    environment.build.version,
+    application.display_version,
+    environment.build.build_id,
+    metadata.uri.app_name,
+    environment.system.os.name AS os,
+    environment.system.os.version AS os_version,
+    environment.build.architecture,
+    normalized_country_code AS country,
+    -- 0 for values retrieved from crash pings
+    0 AS main_crash,
+    0 AS content_crash,
+    0 AS startup_crash,
+    0 AS content_shutdown_crash,
+    LEAST(GREATEST(payload.info.subsession_length / 3600, 0), 25) AS usage_hours,  -- protect against extreme values
+    COALESCE(udf_keyed_histogram_get_sum(payload.keyed_histograms.subprocess_crashes_with_dump, 'gpu'), 0) AS gpu_crashes,
+    COALESCE(udf_keyed_histogram_get_sum(payload.keyed_histograms.subprocess_crashes_with_dump, 'plugin'), 0) AS plugin_crashes,
+    COALESCE(udf_keyed_histogram_get_sum(payload.keyed_histograms.subprocess_crashes_with_dump, 'gmplugin'), 0) AS gmplugin_crashes
+  FROM
+    main_pings
+),
+-- Get mobile usage hours from core pings
+core_ping_data AS (
+  SELECT
+    submission_timestamp,
+    normalized_channel AS channel,
+    metadata.uri.app_version AS version,
+    COALESCE(display_version, metadata.uri.app_version) AS display_version,
+    metadata.uri.app_build_id AS build_id,
+    metadata.uri.app_name,
+    os,
+    osversion AS os_version,
+    arch AS architecture,
+    normalized_country_code AS country,
+    -- 0 for values retrieved from crash pings
+    0 AS main_crash,
+    0 AS content_crash,
+    0 AS startup_crash,
+    0 AS content_shutdown_crash,
+    LEAST(GREATEST(durations / 3600, 0), 25) AS usage_hours,  -- protect against extreme values
```

Comment:
should we be limiting to 24 hours of usage in a day/ping? cc @wlach 
```suggestion
    LEAST(GREATEST(durations / 3600, 0), 24) AS usage_hours,  -- protect against extreme values
```
and should we be making extreme values `NULL` instead? 
```suggestion
    IF(durations/3600 BETWEEN 0 AND 24, durations/3600, NULL) AS usage_hours,  -- protect against extreme values
```

Reply:
I think this should select `payload.info.subsession_length` and `durations` as `usage_seconds`, and then convert to hours and protect against extreme values in the final select, to reduce code duplication

Reply:
Using 24 hours makes sense.  I just copied 25 from the spark job maybe there's a reason for that @wlach knows.  Setting extreme values to `NULL` will inflate the crash rates since they won't contribute to the total usage hours but will contribute crash counts so I'm not sure that's the best way to do it.  Of course setting artificial limits is also not ideal, we just need to accept that the data is somewhat messed up.

Reply:
I just ran a query on main pings and it looks like there are a bunch of subsession lengths that are 24 hours + a bit so maybe that's why.

Reply:
> Setting extreme values to `NULL` will inflate the crash rates since they won't contribute to the total usage hours

setting extreme values to `0` has the same effect on `SUM` as setting them to `NULL`, except that if *all* the values for a group are `NULL` them `SUM` returns `NULL`

Reply:
I was more concerned about the high values than negative.  In a quick scan I didn't find any negative values but a lot of >24 hour values.  Setting those to `NULL` instead of 24 (or 25) will decrease the summed usage hours by a lot.

Reply:
ah, now I understand. You're right, we shouldn't use `NULL`.

Reply:
In general going forward we should use active_ticks for getting usage information, which isn't susceptible to these weird values. The model used in missioncontrol-v2 does this, among other things.

For this dataset however, our priority should be on not doing anything too surprising from the point of view of the users of the mission control dashboard (i.e. try to keep the rate numbers the same, more or less), unless we encounter some kind of egregious error. We know for a fact that this "error aggregates" approach has a *lot* of problems -- that's why we're moving away from it! :smiley: